driver/src/price_estimation/mod.rs

```diff
@@ -0,0 +1,456 @@
+        {
+            prices
+                .entry(token.id)
+                .or_insert_with(|| token.get_price(1.0));
+        }
+
+        PriceEstimates::new(prices)
+    }
+}
+
+/// An opaque token ID.
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
+pub struct TokenId(u16);
+
+/// A token reprensentation.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Clone, Debug)]
+struct Token {
+    id: TokenId,
+    address: Address,
+    symbol: String,
+    decimals: u8,
+}
+
+lazy_static! {
+    static ref SYMBOL_OVERRIDES: HashMap<String, String> = hash_map! {
+        "WETH" => "ETH".to_owned(),
+    };
+}
+
+impl Token {
+    /// Converts the prices from USD into the unit expected by the contract.
+    /// This price is relative to the OWL token which is considered pegged at
+    /// exactly 1 USD with 18 decimals.
+    fn get_price(&self, usd_price: f64) -> u128 {
+        let pow = 36 - (self.decimals as i32);
+        (usd_price * 10.0f64.powi(pow)) as _
+    }
+
+    /// Creates a new token with a fictional address for testing.
+    #[cfg(test)]
+    fn test(index: u16, symbol: &str, decimals: u8) -> Token {
+        Token {
+            id: TokenId(index),
+            address: Address::repeat_byte(index as _),
+            symbol: symbol.into(),
+            decimals,
+        }
+    }
+}
+
+impl Serialize for TokenId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        format!("T{:04}", self.0).serialize(serializer)
+    }
+}
+
+/// A reader for retrieving ERC20 token information.
+#[cfg_attr(test, mockall::automock)]
+trait TokenReader {
+    /// Reads a token given its index in the exchange contract.
+    fn read_token(&self, index: u16) -> Result<Token>;
+}
+
+/// Default token reader implementation.
+struct EthTokenReader {
+    contract: BatchExchange,
+}
+
+impl EthTokenReader {
+    /// Creates a new token reader from a web3 provider and exchange contract
+    /// instance.
+    fn new(contract: &BatchExchange) -> Self {
+        EthTokenReader {
+            contract: contract.clone(),
+        }
+    }
+}
+
+impl TokenReader for EthTokenReader {
+    /// Retrieves the token information from the Ethereum network.
+    fn read_token(&self, index: u16) -> Result<Token> {
+        let address = self
+            .contract
+            .token_id_to_address_map(index as _)
+            .call()
+            .wait()
+            .unwrap();
+        let erc20 = ERC20Detailed::at(&self.contract.web3(), address);
+        let symbol = {
+            // NOTE: We check if the symbol is part of the overrides map, and
+            //   use the overridden value if it is. This allows ERC20 tokens
+            //   like WETH to be treated by ETH, since exchanges generally only
+            //   track ETH prices and not WETH.
+            let symbol = erc20.symbol().call().wait().unwrap();
```

Comment:
I wonder if it is save to rely on the ERC20 symbol. I can create my own contract with an arbitrary symbol. Personally I would feel more comfortable using e.g. our trusted [token list](https://github.com/gnosis/dex-js/blob/master/src/tokenList.json) to get the symbols for tokens we now are real.

But then, all you can really do is get wrong estimate for your fake token, not sure if that is a concern.

Reply:
I figured that we would eventually extend token filtering to this here.

Reply:
I could imaging that we might not want to fully blacklist a token but still not necessarily do a price estimation for them.

Reply:
We can implement a price estimation token filter analogous to the orderbook filter. This would allow us to quickly filter tokens without needing to rebuild. Or, for price estimation, we can go the opposite direction and have a whitelist (instead of a blacklist).